stdlib/src/mmap.rs

```diff
@@ -0,0 +1,1119 @@
+            );
+
+            Ok(buf)
+        }
+    }
+
+    impl AsMapping for PyMmap {
+        const AS_MAPPING: PyMappingMethods = PyMappingMethods {
+            length: Some(|mapping, _vm| Ok(Self::mapping_downcast(mapping).len())),
+            subscript: Some(|mapping, needle, vm| {
+                Self::mapping_downcast(mapping)._getitem(needle, vm)
+            }),
+            ass_subscript: Some(|mapping, needle, value, vm| {
+                let zelf = Self::mapping_downcast(mapping);
+                if let Some(value) = value {
+                    Self::_setitem(zelf.to_owned(), needle, value, vm)
+                } else {
+                    Err(vm.new_type_error("mmap object doesn't support item deletion".to_owned()))
+                }
+            }),
+        };
+    }
+
+    impl AsSequence for PyMmap {
+        const AS_SEQUENCE: PySequenceMethods = PySequenceMethods {
+            length: Some(|seq, _vm| Ok(Self::sequence_downcast(seq).len())),
+            item: Some(|seq, i, vm| {
+                let zelf = Self::sequence_downcast(seq);
+                zelf.get_item_by_index(i, vm)
+            }),
+            ass_item: Some(|seq, i, value, vm| {
+                let zelf = Self::sequence_downcast(seq);
+                if let Some(value) = value {
+                    Self::set_item_by_index(zelf.to_owned(), i, value, vm)
+                } else {
+                    Err(vm.new_type_error("mmap object doesn't support item deletion".to_owned()))
+                }
+            }),
+            ..PySequenceMethods::NOT_IMPLEMENTED
+        };
+    }
+
+    #[pyimpl(with(Constructor, AsMapping, AsSequence, AsBuffer), flags(BASETYPE))]
+    impl PyMmap {
+        fn as_bytes_mut(&self) -> BorrowedValueMut<[u8]> {
+            PyMutexGuard::map(self.mmap.lock(), |m| match m {
+                MmapObj::Read(_) => panic!("mmap can't modify a readonly memory map."),
+                MmapObj::Write(mmap) => &mut mmap[..],
+                _ => panic!("mmap closed or invalid"),
+            })
+            .into()
+        }
+
+        fn as_bytes(&self) -> BorrowedValue<[u8]> {
+            PyMutexGuard::map_immutable(self.mmap.lock(), |m| match m {
+                MmapObj::Read(mmap) => &mmap[..],
+                MmapObj::Write(mmap) => &mmap[..],
+                _ => panic!("mmap closed or invalid"),
+            })
+            .into()
+        }
+
+        #[pymethod(magic)]
+        pub(crate) fn len(&self) -> usize {
+            self.inner_size() as usize
+        }
+
+        #[inline]
+        fn inner_size(&self) -> isize {
+            self.size.load()
+        }
+
+        #[inline]
+        fn inner_pos(&self) -> isize {
+            self.pos.load()
+        }
+
+        #[inline]
+        fn advance_pos(&self, step: isize) {
+            self.pos.store(self.inner_pos() + step);
+        }
+
+        #[inline]
+        fn check_writable(&self, vm: &VirtualMachine) -> PyResult<()> {
```

Comment:
Right now it dosen't need to return anything, just a validate funciton.

Reply:
Yes, checking itself doesn't need return value. I will describe it more.

Let's see `setitem_by_index` method. The 'checked' status only exists in human recognition for now.

```rust 
            zelf.check_writable(vm)?;  // we checked writable here
            let b = value_from_object(vm, &value)?;

            match zelf.check_valid(vm)?.deref_mut().as_mut().unwrap() {
                MmapObj::Write(mmap) => mmap[i as usize] = b,
                _ => unreachable!("already check"),  // but it didn't propagate to here
            };
```

Instead of it, we can let code knows it.

```rust
            let mmap = zelf.borrow_writable(vm)?;
            let b = value_from_object(vm, &value)?;
            mmap[i as usize] = b;
```

When returning borrow is not very easy, we do it in closure

```rust
            zelf.try_writable(|mmap| {
                let b = value_from_object(vm, &value)?;
                mmap[i as usize] = b;
            }, vm)?;
```

Reply:
I refactor this in https://github.com/RustPython/RustPython/pull/3755/commits/2c8ef17c3609b7b60b7f0d1f884e2aa8bf66cf4d .

It's more clear, thank you.